docs/design/namespaces.md

```diff
@@ -1,193 +1,219 @@
+type NamespaceStatus struct {
+  // Phase is the current lifecycle phase of the namespace.
+  Phase NamespacePhase `json:"phase,omitempty"`
+  // Finalizers is the list of named finalize actions that have completed
+  Finalizers []FinalizerName `json:"finalizers"`
+}
+```
 
-In etcd, we want to continue to still support efficient WATCH across namespaces.
+A *FinalizerName* is a qualified name.
 
-Resources that persist content in etcd will have storage paths as follows:
+A *Finalizer* denotes a named entity that must release it's claim to the *Namespace* resource
+prior to its deletion.
 
-/registry/{resourceType}/{resource.Namespace}/{resource.Name} 
+The *Finalizer* pattern allows for asynchronous controllers that are watching cluster state
+to take part in distributed graceful termination of *Namespace* managed resources that are
+unknown to the cluster.
 
-This enables k8s service to WATCH /registry/{resourceType} for changes across namespace of a particular {resourceType}.
+During *Namespace* termination, each *Finalizer* must perform a *finalize* operation before
+the cluster is permitted to delete the *Namespace*.
 
-Upon scheduling a pod to a particular host, the pod's namespace must be in the key path as follows:
+A *finalize* operation updates the *NamespaceStatus.Finalizers* list noting that the named
+finalizer has completed its operation.
 
-/host/{host}/pod/{pod.Namespace}/{pod.Name}
+### Phases
 
-## k8s Authorization service
+A *Namespace* may exist in the following phases.
 
-This design assumes the existence of an authorization service that filters incoming requests to the k8s API Server in order
-to enforce user authorization to a particular k8s resource.  It performs this action by associating the *subject* of a request
-with a *policy* to an associated HTTP path and verb.  This design encodes the *namespace* in the resource path in order to enable
-external policy servers to function by resource path alone.  If a request is made by an identity that is not allowed by 
-policy to the resource, the request is terminated.  Otherwise, it is forwarded to the apiserver.
+```
+type NamespacePhase string
+const(
+  NamespaceActive NamespacePhase = "Active"
+  NamespaceTerminating NamespaceTerminating = "Terminating"
+)
+
+type NamespaceStatus struct { 
+  ...
+  Phase NamespacePhase 
+}
+```
 
-## k8s controller-manager
+**Active**
 
-The controller-manager will provision pods in the same namespace as the associated replicationController.
+Upon creation, a *Namespace* goes in the *Active* phase.  This means that content may be associated with
+a namespace, and all normal interactions with the namespace are allowed to occur in the cluster.
 
-## k8s Kubelet
+DELETE *Namespace* to move to *Terminating* phase
 
-There is no major change to the kubelet introduced by this proposal.
+**Terminating**
 
-### kubecfg client
+Admission control prevents any new content from being in that *Namespace*.
```

Comment:
The reason to enforce via admission control is as follows:

When a ```NamespaceController``` sees a amespace object whose  ```Namespace.Status.Phase``` is ```Terminating```, it will enumerate each known resource type in that namespace and delete it one by one.  Admission control blocks creation of new resources in that namespace in order to prevent a race-condition where the controller could believe all of a given resource type had been deleted from the namespace, when in fact some other rogue client agent had created new ones.